components/tools/OmeroPy/src/omero/gateway/__init__.py

```diff
@@ -9032,7 +9032,13 @@ def renderJpegRegion(self, z, t, x, y, width, height, level=None,
         self._pd.region = regionDef
         try:
             if level is not None:
-                self._re.setResolutionLevel(level)
+                if self._re.requiresPixelsPyramid():
+                    self._re.setResolutionLevel(level)
+                else:
+                    if level != 0:
```

Comment:
This feels like it should be a hard error. Probably an HTTP 400.

Reply:
The method will return ``None`` and that will result in a ``404`` in views.py

Reply:
Yes, which is not really an appropriate error in this case.

Reply:
I read 404 in your comment

Reply:
in that case the check will need a check too in views.py so it does not change the existing behaviour i.e. none => 404
that will have an impact on performance since we will have to handle web usage and gateway usage

Reply:
Just looking at the overarching semantics, it looks like we have always been catching the `omero.InternalException` and returning `None`. While I don't really agree with that decision, it has been the prevailing semantic for some time now and I don't think it's a good idea to change that. Checking the level value preemptively is not really something you need to do here just for the sake of avoiding an exception.

Incidentally, these exceptions should probably be logged at `WARN` rather than `DEBUG` so that we actually see them in the logs.

Reply:
With the omero.InternalException handling, ```conn.renderJpegRegion()``` will also return ```None``` if you specify an invalid ```?tile=3,5,100,256,256``` (tile outside of image) for a tiled image, so having it return ```None``` for an invalid ```level``` (for a non-tiled image) seems consistent.

This hides the inconsistency on the server where the first will throw a ```ome.conditions.InternalException:  Wrapped Exception: (java.lang.RuntimeException):
Invalid Region, the height must be positive:0``` and the second will simply render something other than what you might expect.

Now that @waxenegger has reverted to using ```?region= ``` for large non-tiled images (https://github.com/ome/omero-iviewer/pull/186/commits/71d9f0d7e23a8dfd5e8cd3abf8a2d9a378e13fc6) this is no-longer so urgent.
Seems we can't move these changes to the server for a point release since it's an API change, but I would have thought that the changes in this PR are a good first step?

Reply:
Outwith this PR, do feel free to open a design issue that accumulates suggested server exception changes that would improve the API usability - maybe we can try to take on a bunch next year or whenever next time we're breaking everything.

Reply:
The fix is not bound to iviewer.
There are few usage issues with the methods.
As @chris-allan pointed out, more check => potential performance issue